File: day4/4.hs
import Data.Char (isDigit)

nums :: String -> [Int]
nums [] = []
nums xs = if null n then [] else read n : nums tl
  where
    hd = dropWhile (not . isDigit) xs
    n = takeWhile isDigit hd
    tl = dropWhile isDigit hd

readLines :: FilePath -> IO [String]
readLines file = do
  contents <- readFile file
  return $ lines contents

wins :: String -> Int
wins game = length (take 10 n `intersect` drop 10 n)
  where
    n = tail $ nums game
    intersect [] = const []
    intersect xs = filter (`elem` xs)

p1 :: [String] -> Int
p1 games = sum $ map (ep . wins) games
  where ep x = if x > 0 then 2 ^ (x - 1) else 0

p2 :: [String] -> [Int]
p2 [] = []
p2 (x : xs) = (1 + sum (take (wins x) r)) : r
  where r = p2 xs

main :: IO ()
main = do
  games <- readLines "input.txt"
  print $ p1 games
  print $ sum $ p2 games

File: day4/4.c

#include "../a.h"

I g[209][36], sc[209], cp[209];

I main(V) {
    FILE*f=fopen("input.txt", "r");
    C*p,buf[1024];I p1s=0,p2s=0;
    for(I i=0,j;fgets(buf,1024,f);i++) {
        j=0,p=buf;W(*p){W(!isdigit(*p)&&*p)p++;g[i][j++]=atoi(p);W(isdigit(*p)&&*p)p++;}}
    Fi(209,I c[100]={0};Fj(35,c[g[i][j+1]]++)Fj(100,c[j]==2&&sc[i]++)cp[i]=1)
    Fi(209,p1s+=sc[i]==0?0:1<<sc[i]-1;Fj(sc[i],j+i<208&&(cp[j+i+1]+=cp[i])))
    Fi(209,p2s+=cp[i])
    printf("%d\n%d\n",p1s,p2s);
}

